exchanges/poloniex/poloniex_websocket_trade.go

```diff
@@ -0,0 +1,98 @@
+)
+
+// WsCreateOrder create an order for an account.
+func (e *Exchange) WsCreateOrder(ctx context.Context, arg *PlaceOrderRequest) (*OrderIDResponse, error) {
+	if err := validateOrderRequest(arg); err != nil {
+		return nil, err
+	}
+	var resp []*OrderIDResponse
+	if err := e.SendWebsocketRequest(ctx, "createOrder", arg, &resp); err != nil {
+		return nil, fmt.Errorf("%w: %w", order.ErrPlaceFailed, err)
+	}
+	if len(resp) != 1 {
+		return nil, fmt.Errorf("%w: %w", order.ErrPlaceFailed, common.ErrInvalidResponse)
+	} else if resp[0].Code != 0 && resp[0].Code != 200 {
+		return nil, fmt.Errorf("%w: error code: %d message: %s", order.ErrPlaceFailed, resp[0].Code, resp[0].Message)
+	}
+	return resp[0], nil
+}
+
+// WsCancelMultipleOrdersByIDs batch cancel one or many active orders in an account by IDs through the websocket stream.
+func (e *Exchange) WsCancelMultipleOrdersByIDs(ctx context.Context, orderIDs, clientOrderIDs []string) ([]*WsCancelOrderResponse, error) {
+	if len(clientOrderIDs) == 0 && len(orderIDs) == 0 {
+		return nil, order.ErrOrderIDNotSet
+	}
+	params := make(map[string][]string)
+	if len(clientOrderIDs) > 0 {
+		params["clientOrderIds"] = clientOrderIDs
+	}
+	if len(orderIDs) > 0 {
+		params["orderIds"] = orderIDs
+	}
+	var resp []*WsCancelOrderResponse
+	if err := e.SendWebsocketRequest(ctx, "cancelOrders", params, &resp); err != nil {
+		return nil, fmt.Errorf("%w: %w", order.ErrCancelFailed, err)
+	}
+	return resp, nil
+}
+
+// WsCancelTradeOrders batch cancel all orders in an account.
+func (e *Exchange) WsCancelTradeOrders(ctx context.Context, symbols []string, accountTypes []AccountType) ([]*WsCancelOrderResponse, error) {
+	args := make(map[string]any)
+	if len(symbols) > 0 {
+		args["symbols"] = symbols
+	}
+	if len(accountTypes) > 0 {
+		args["accountTypes"] = accountTypes
+	}
+	var resp []*WsCancelOrderResponse
+	if err := e.SendWebsocketRequest(ctx, "cancelAllOrders", args, &resp); err != nil {
+		return nil, fmt.Errorf("%w: %w", order.ErrCancelFailed, err)
+	}
+	return resp, nil
+}
+
+// SendWebsocketRequest sends a websocket request through the private connection
+func (e *Exchange) SendWebsocketRequest(ctx context.Context, event string, arg, response any) error {
+	conn, err := e.Websocket.GetConnection(connSpotPrivate)
+	if err != nil {
+		return err
+	}
+	input := &struct {
+		ID     string `json:"id"`
+		Event  string `json:"event"`
+		Params any    `json:"params"`
+	}{
+		ID:     e.MessageID(),
+		Event:  event,
+		Params: arg,
+	}
+	e.spotSubMtx.Lock()
+	result, err := conn.SendMessageReturnResponse(ctx, sWebsocketPrivateEPL, input.ID, input)
+	e.spotSubMtx.Unlock()
+	if err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(result, &WebsocketResponse{Data: response}); err != nil {
+		return err
+	}
+	if response == nil {
+		return common.ErrNoResponse
+	}
+	if errTypes, ok := response.([]hasError); ok {
+		for _, errType := range errTypes {
+			err = common.AppendError(err, errType.Error())
+		}
+	}
```

Comment:
🚧 Change request:
Sorry / Not Sorry: This does not appear to work.
Did you test it at all ?

Go does not seem to support casting `any` to a slice of interfaces, including `[]any`.
My testing showed you can only cast it to the concrete type it actually is.

Sucks - Sorry for leading you down this path. I should have tested that patch for you.

So I think you should move the Error() calls up into the direct callers of this, which would be the most abstract version.

And at the same time, you can add the context that wrappers have been adding.

Reply:
Well, well, well ... I have altered this and tested the new implementation.  Please check it out.